opm/grid/cpgrid/CpGrid.cpp

```diff
@@ -1471,6 +1471,490 @@ template cpgrid::Entity<0> createEntity(const CpGrid&, int, bool);
+        const int refinedLevelGridIdx = level + preAdaptMaxLevel +1;
+#if HAVE_MPI
+        refined_grid_ptr_vec[level] = std::make_shared<Dune::cpgrid::CpGridData>((*(this-> data_[0])).ccobj_, refined_data_vec[level]);
+#else
+        // DUNE 2.7 is missing convertion to NO_COMM
+        refined_grid_ptr_vec[level] = std::make_shared<Dune::cpgrid::CpGridData>(refined_data_vec[level]);
+#endif
+        // Store refined grid
+        (this-> data_).push_back(refined_grid_ptr_vec[level]);
+
+        Dune::cpgrid::DefaultGeometryPolicy&  refinedLevel_geometries = (*data_[refinedLevelGridIdx]).geometry_;
+        // Mutable containers for adapted corners, faces, cells, face tags, and face normals.
+        Dune::cpgrid::EntityVariableBase<cpgrid::Geometry<0,3>>& level_corners =
+            *(refinedLevel_geometries.geomVector(std::integral_constant<int,3>()));
+        Dune::cpgrid::EntityVariableBase<cpgrid::Geometry<2,3>>& level_faces =
+            *(refinedLevel_geometries.geomVector(std::integral_constant<int,1>()));
+        Dune::cpgrid::EntityVariableBase<cpgrid::Geometry<3,3>>& level_cells =
+            *(refinedLevel_geometries.geomVector(std::integral_constant<int,0>()));
+
+        level_corners = refined_corners_vec[level];
+        level_faces = refined_faces_vec[level];
+        level_cells = refined_cells_vec[level];
+
+        (*data_[refinedLevelGridIdx]).cell_to_point_ = refined_cell_to_point_vec[level];
+        (*data_[refinedLevelGridIdx]).cell_to_face_ = refined_cell_to_face_vec[level];
+
+        (*data_[refinedLevelGridIdx]).face_to_point_ = refined_face_to_point_vec[level];
+        (*data_[refinedLevelGridIdx]).face_to_cell_ = refined_face_to_cell_vec[level];
+
+        cpgrid::EntityVariable<enum face_tag,1>& level_face_tags =   (*data_[refinedLevelGridIdx]).face_tag_;
+        Dune::cpgrid::EntityVariableBase<enum face_tag>& level_mutable_face_tags = level_face_tags;
+        level_mutable_face_tags = mutable_refined_face_tags_vec[level];
+
+        cpgrid::SignedEntityVariable<Dune::FieldVector<double,3>,1>&  level_face_normals =   (*data_[refinedLevelGridIdx]).face_normals_;
+        Dune::cpgrid::EntityVariableBase<PointType>& level_mutable_face_normals = level_face_normals;
+        level_mutable_face_normals = mutable_refined_face_normals_vec[level];
+
+        // Further Refined grid Attributes
+        //
+        // Populate some attributes of the level LGR
+        (*data_[refinedLevelGridIdx]).level_data_ptr_ = &(this -> data_);
+        (*data_[refinedLevelGridIdx]).level_ = refinedLevelGridIdx;
+        this -> lgr_names_[lgr_name_vec[level]] = refinedLevelGridIdx; // {"name_lgr", level}
+        (*data_[refinedLevelGridIdx]).child_to_parent_cells_ = refined_child_to_parent_cells_vec[level];
+        (*data_[refinedLevelGridIdx]).cell_to_idxInParentCell_ = refined_cell_to_idxInParentCell_vec[level];
+        (*data_[refinedLevelGridIdx]).level_to_leaf_cells_ =  refined_level_to_leaf_cells_vec[level];
```

Comment:
I have the feeling this here leads to dangling pointers in the Geometry instances that you have populated in all those helper functions above. The pointers in there will still point to the temporary containers refined*, but those containers will be destroyed eventually.

Would it be possible to already create std::shared_ptr<CpGridData> for all the levels and the new leaf before caling those functions. THen we can pass references to the containers in there to all these functions to populate them directly?

Here is a rough sketch of my idea:
```
std::shared_ptr<CpGridData> level3_data = {};
std::vector<std::reference_wrapper<Dune::cpgrid::EntityVariableBase<cpgrid::Geometry<0,3>>>> refined_corners_vec(levels);
refined_corners_vec[3] = *(refinedLevel_geometries.geomVector(std::integral_constant<int,3>()));
...
// populate all the containes ...
....
```

The alternative would be to correct all the pointers in the Geometry instances after copying to CpGridData members here. I think both is OK. Just keep in mind that the copies also cost time and memory.

Reply:
I think there is an easier and more elegant solution:

Use std::swap or the swap member function instead of assignment. 
```
(*data_[refinedLevelGridIdx]).child_to_parent_cells_.swap(refined_child_to_parent_cells_vec[level]);
...
```
That is supposed to simply exchange the underlying pointers, which is exactly what we want here.  For this we might need implement swap member functions for `SparseTable`, that calls swap on all members, and `OrientedEntityTable` would simply hve `using super_t::swap;`.

Reply:
Great catch! Many thanks. Now it fixed. most of the attributes now get swapped.  swap added in EntitityVaraibleBase and OrientedEntityTable.